src/java/org/apache/cassandra/locator/Ec2MetadataServiceConnector.java

```diff
@@ -0,0 +1,234 @@
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.locator;
+
+import java.io.IOException;
+import java.time.Duration;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableMap;
+
+import org.apache.cassandra.exceptions.ConfigurationException;
+import org.apache.cassandra.utils.Clock;
+import org.apache.cassandra.utils.Pair;
+
+import static java.lang.String.format;
+
+public interface Ec2MetadataServiceConnector extends CloudApiCallable
+{
+    String EC2_METADATA_TYPE_PROPERTY = "ec2_metadata_type";
+    String DEFAULT_METADATA_SERVICE_URL = "http://169.254.169.254";
+
+    enum EC2MetadataType
+    {
+        V1("v1"),
+        V2("v2");
+
+        private final String name;
+
+        EC2MetadataType(String name)
+        {
+            this.name = name;
+        }
+
+        @Override
+        public String toString()
+        {
+            return name;
+        }
+    }
+
+    static Ec2MetadataServiceConnector create(SnitchProperties props)
+    {
+        try
+        {
+            EC2MetadataType type = EC2MetadataType.valueOf(props.get(EC2_METADATA_TYPE_PROPERTY, EC2MetadataType.V2.toString()).toUpperCase());
+
+            switch (type)
+            {
+                case V1:
+                    return V1Connector.create(props);
+                case V2:
+                    return V2Connector.create(props);
+                default:
+                    throw new IllegalArgumentException();
+            }
+        }
+        catch (IllegalArgumentException ex)
+        {
+            throw new ConfigurationException(format("%s must be one of %s", EC2_METADATA_TYPE_PROPERTY,
+                                                    Arrays.stream(EC2MetadataType.values())
+                                                          .map(EC2MetadataType::toString)
+                                                          .collect(Collectors.joining(", "))));
+        }
+    }
+
+    class V1Connector implements Ec2MetadataServiceConnector
+    {
+        @VisibleForTesting
+        static final String EC2_METADATA_URL_PROPERTY = "ec2_metadata_url";
+
+        private final String metadataServiceUrl;
+
+        static V1Connector create(SnitchProperties props)
+        {
+            String url = props.get(EC2_METADATA_URL_PROPERTY, DEFAULT_METADATA_SERVICE_URL);
+            return new V1Connector(url);
+        }
+
+        public V1Connector(String metadataServiceUrl)
+        {
+            this.metadataServiceUrl = metadataServiceUrl;
+        }
+
+        @Override
+        public String getApiUrl()
+        {
+            return metadataServiceUrl;
+        }
+
+        @Override
+        public String toString()
+        {
+            return "V1Connector{" + EC2_METADATA_URL_PROPERTY + '=' + getApiUrl() + '}';
```

Comment:
probably we should return full name, that is, including the information that it is EC2 connector